src/query/shorthand.ts

```diff
@@ -277,3 +282,205 @@ export function fieldDef(encQ: EncodingQuery,
   }
   return fieldAndParams;
 }
+
+const CHANNEL_INDEX = toMap(CHANNELS);
+const AGGREGATE_OP_INDEX = toMap(AGGREGATE_OPS);
+const SINGLE_TIMEUNIT_INDEX = toMap(SINGLE_TIMEUNITS);
+const MULTI_TIMEUNIT_INDEX = toMap(MULTI_TIMEUNITS);
+
+export function parse(shorthand: string): SpecQuery {
+  // TODO(https://github.com/uwdata/compassql/issues/259):
+  // Do not split directly, but use an upgraded version of `getClosingBraceIndex()`
+  let splitShorthand = shorthand.split('|');
+
+  let specQ: SpecQuery = {mark: splitShorthand[0], encodings: [] as EncodingQuery[]};
+
+  for (let i = 1; i < splitShorthand.length; i++) {
+    let part = splitShorthand[i];
+    const splitPart = splitWithTail(part, ':', 1);
+    const splitPartKey = splitPart[0];
+    const splitPartValue = splitPart[1];
+
+    if (CHANNEL_INDEX[splitPartKey] || splitPartKey === '?') {
+      shorthandParser.encoding(specQ, splitPartKey, splitPartValue);
```

Comment:
Since you're not using output from shorthandParser.encoding to determine whether you can skip the rest of stuff, I wonder why you have to pass `specQ` inside?

You can just return `encQ` and do 

```
const encQ =  encoding(specQ, splitPartKey, splitPartValue)
specQ.encodings.push(encQ);
```

This way, your test code will be way better abstracted.